src/Microsoft.TestPlatform.Common/Utilities/TypesToLoadUtilities.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright(c) Microsoft Corporation.All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.Common.Utilities
+{
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Utilities;
+
+    using System;
+    using System.Collections.Generic;
+    using System.Reflection;
+    using System.Linq;
+
+    internal static class TypesToLoadUtilities
+    {
+        public const string TypesToLoadAttributeFullName = "Microsoft.VisualStudio.TestPlatform.ObjectModel.Utilities.TypesToLoadAttribute";
+
+        internal static IEnumerable<Type> GetTypesToLoad(Assembly assembly)
+        {
+            ValidateArg.NotNull(assembly, nameof(assembly));
+
+            var typesToLoad = assembly
+                .GetCustomAttribute(TypesToLoadAttributeFullName)
+                .SelectMany(i => GetTypesToLoad(i));
+
+            return typesToLoad;
+        }
+
+        private static IEnumerable<Type> GetTypesToLoad(Attribute attribute)
+        {
+            if (attribute == null)
+                return Enumerable.Empty<Type>();
+
+            var type = attribute.GetType();
+            var typesProperty = type.GetProperty("Types");
```

Comment:
This method contains reflection, which is slow. I assume how much improvement it could give us with having custom attribute?

Reply:
The other implementation was also using reflection. 

Our performance gain is because this will save us from scanning the entire assembly.

Reply:
In my change we did not have reflection, the logic before it used reflection. It used `GetTypes()` on assembly, but we changed to load only specific types. But now you included that logic in each type, where for each assembly you again call `GetType`.
So I suspicious in overall benefit, which it could give us.

Reply:
`GetCustomAttribute<T>` does use reflection, as this is. Only extra is now, instead of matching the attribute by type, it's matching by name. This is a necessary evil to keep this implementation.